google-cloud-firestore/src/main/java/com/google/cloud/firestore/StreamableQuery.java

```diff
@@ -0,0 +1,400 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore;
+
+import static com.google.cloud.firestore.telemetry.TraceUtil.*;
+import static com.google.common.collect.Lists.reverse;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.SettableApiFuture;
+import com.google.api.gax.rpc.ApiStreamObserver;
+import com.google.api.gax.rpc.ResponseObserver;
+import com.google.api.gax.rpc.StatusCode;
+import com.google.api.gax.rpc.StreamController;
+import com.google.cloud.Timestamp;
+import com.google.cloud.firestore.telemetry.TraceUtil;
+import com.google.cloud.firestore.v1.FirestoreSettings;
+import com.google.common.collect.ImmutableMap;
+import com.google.firestore.v1.Document;
+import com.google.firestore.v1.RunQueryRequest;
+import com.google.firestore.v1.RunQueryResponse;
+import com.google.protobuf.ByteString;
+import io.grpc.Status;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.threeten.bp.Duration;
+
+/**
+ * Represents a query whose results can be streamed. If the stream fails with a retryable error,
+ * implementations of StreamableQuery can optionally support retries with a cursor, as indicated by
+ * `isRetryableWithCursor`. Retrying with a cursor means that the StreamableQuery can be resumed
+ * where it failed by first calling `startAfter(lastDocumentReceived)`.
+ */
+public abstract class StreamableQuery<SnapshotType> {
+  final Query.QueryOptions options;
+  final FirestoreRpcContext<?> rpcContext;
+
+  StreamableQuery(FirestoreRpcContext<?> rpcContext, Query.QueryOptions options) {
+    this.rpcContext = rpcContext;
+    this.options = options;
+  }
+
+  abstract RunQueryRequest.Builder toRunQueryRequestBuilder(
+      @Nullable final ByteString transactionId,
+      @Nullable final Timestamp readTime,
+      @Nullable ExplainOptions explainOptions);
+
+  abstract boolean isRetryableWithCursor();
+
+  abstract StreamableQuery startAfter(@Nonnull DocumentSnapshot snapshot);
```

Comment:
I realize this is pre-existing, but it really should be fixed.

Also, the `Query` class should have it's `startAfter` method marked with `@Override`

```suggestion
  abstract StreamableQuery<QuerySnapshot> startAfter(@Nonnull DocumentSnapshot snapshot);
```

Reply:
Updated. Thanks.